pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test: TestCCLLogic_regional_by_row_hash_sharded_index failed

pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test.TestCCLLogic_regional_by_row_hash_sharded_index [failed](https://teamcity.cockroachdb.com/buildConfiguration/BUILDTYPE_ID-not-found-in-env/6369522?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/BUILDTYPE_ID-not-found-in-env/6369522?buildTab=artifacts#/) on master @ [2372698da1dfacb90f60c6a63f2c1298d1db16b8](https://github.com/cockroachdb/cockroach/commits/2372698da1dfacb90f60c6a63f2c1298d1db16b8):


```
=== RUN   TestCCLLogic_regional_by_row_hash_sharded_index
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/231e1c2b82a8c60573009e2061c3fb3c/logTestCCLLogic_regional_by_row_hash_sharded_index2383282262
    test_log_scope.go:79: use -show-logs to present logs inline
[05:18:47] setting distsql_workmem='76905B';
[05:18:47] rng seed: -416924996751357404

[05:40:58] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3884/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test_/multiregion-9node-3region-3azs-no-los_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index: 11 statements
[06:03:31] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3884/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test_/multiregion-9node-3region-3azs-no-los_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index: 13 statements
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCCLLogic_regional_by_row_hash_sharded_index.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-19362

[Link](https://teamcity.cockroachdb.com/repository/download/Cockroach_ScratchProjectPutTcExperimentsInHere_SqlLogicDeclarativeSchemaChangerCorpusBazel/6369522:id/bazel-testlogs/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test/shard_16_of_16/test.log)
```
=== RUN   TestCCLLogic_regional_by_row_hash_sharded_index
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/231e1c2b82a8c60573009e2061c3fb3c/logTestCCLLogic_regional_by_row_hash_sharded_index2383282262
    test_log_scope.go:79: use -show-logs to present logs inline
*
* INFO: Running test with the default test tenant. If you are only seeing a test case failure when this message appears, there may be a problem with your test case running within tenants.
*
[05:18:47] setting distsql_workmem='76905B';
[05:18:47] rng seed: -416924996751357404

[05:40:58] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3884/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test_/multiregion-9node-3region-3azs-no-los_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index: 11 statements
[06:03:31] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3884/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/multiregion-9node-3region-3azs-no-los/multiregion-9node-3region-3azs-no-los_test_/multiregion-9node-3region-3azs-no-los_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index: 13 statements
panic: test timed out after 59m55s
...
```
If I'm counting correctly, statement 11 is `CREATE INDEX ON t_to_be_hashed (b) USING HASH;` and statement 13 is `CREATE UNIQUE INDEX ON t_to_be_hashed (b) USING HASH;` which probably indicates that schema changes are very slow.

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -6269,8 +6269,9 @@ func ReplacePointTombstonesWithRangeTombstones(
 
 		key := iter.UnsafeKey()
 
-		// Skip intents and inline values.
-		if key.Timestamp.IsEmpty() {
+		// Skip intents and inline values, and system table keys which
+		// might be watched by rangefeeds.
+		if key.Timestamp.IsEmpty() || isWatchedSystemTable(key.Key) {
 			iter.NextKey()
 			continue
 		}
@@ -6361,4 +6362,30 @@ func ReplacePointTombstonesWithRangeTombstones(
 	}
 
 	return nil
+}
+
+// In order to test the correctness of range deletion tombstones, we added a
+// testing knob to replace point deletions with range deletion tombstones in
+// some tests. Unfortuantely, doing so affects the correctness of rangefeeds.
+// The tests in question do not use rangefeeds, but some system functionality
+// does use rangefeeds internally. The primary impact is that catch-up scans
+// will miss deletes. That makes these issues rare and hard to detect. In order
+// to deflake these tests, we avoid rewriting deletes on relevant system
+// tables.
+func isWatchedSystemTable(key roachpb.Key) bool {
+	rem, _, err := keys.DecodeTenantPrefix(key)
+	if err != nil { // allow unprefixed keys to pass through
+		return false
+	}
+	_, tableID, _, err := keys.DecodeTableIDIndexID(rem)
+	if err != nil { // allow keys which do not correspond to sql tables
+		return false
+	}
+	switch tableID {
+	case keys.SettingsTableID, keys.SpanConfigurationsTableID,
+		keys.SQLInstancesTableID, keys.DescriptorTableID, keys.ZonesTableID:
+		return true
+	default:
+		return false
+	}
 }

